graph/articulationpoints.go

```diff
@@ -0,0 +1,100 @@
+package graph
+
+import "github.com/TheAlgorithms/Go/math/min"
+
+type apHelper struct {
+	is_ap              []bool
+	visited            []bool
+	child_cnt          []int
+	discovery_time     []int
+	earliest_discovery []int
+}
+
+// ArticulationPoint is a function to identify articulation points in a graph.
+// The function takes the graph as an argument and returns a boolean slice
+// which indicates whether a vertex is an articulation point or not.
+// Worst Case Time Complexity: O(|V| + |E|)
+// Auxiliary Space: O(|V|)
+// reference: https://en.wikipedia.org/wiki/Biconnected_component and https://cptalks.quora.com/Cut-Vertex-Articulation-point
+func ArticulationPoint(graph Graph) []bool {
+	// time variable to keep track of the time of discovery_time of a vertex
+	time := 0
+
+	//initialize all the variables
+	apHelperInstance := &apHelper{
+		is_ap:     make([]bool, graph.vertices),
+		visited:   make([]bool, graph.vertices),
+		child_cnt: make([]int, graph.vertices),
+		// integer slice to store the discovery time of a vertex as we traverse
+		// the graph in a depth first manner
+		discovery_time: make([]int, graph.vertices),
+		// integer slice to store the earliest discovered vertex reachable from a vertex
+		earliest_discovery: make([]int, graph.vertices),
+	}
+	articulationPointHelper(
+		apHelperInstance,
+		0,
+		-1,
+		&time,
+		graph,
+	)
+
+	if apHelperInstance.child_cnt[0] == 1 {
+		// if the root has only one child, it is not an articulation point
+		apHelperInstance.is_ap[0] = false
+	}
+
+	return apHelperInstance.is_ap
+}
+
+// articulationPointHelper is a recursive function to traverse the graph
+// and mark articulation points. Based on the depth first search transversal
+// of the graph, however modified to keep track and update the
+// `child_cnt`, `discovery_time`` and `earliest_discovery` slices defined above
+func articulationPointHelper(
+	apHelperInstance *apHelper,
+	vertex int,
+	parent int,
+	time *int,
+	graph Graph,
```

Comment:
Yeah I did consider that. However, I decided against it simply because I thought this way its more clear about the usage of the function parameters. The "pure" parameters being passed as values while the "impure" parameters being passed as reference to be changed and tinkered inside the functions forming a "clearer" distinction within the parameters. I do understand that passing the graph variable by reference help squeeze out the last bit of execution optimization but as a normal paradigm I tried to keep the function as "pure" as possible. I don't know if has to do with standard golang conventions, in which case I can modify this to be passed as reference in my implementation.

Reply:
No nothing to do with standard Golang convention, rather its the idea related to pointer operations. If you're not modifying the value of the `object` then pass the pointer instead of a copy. However, if you are modifying the value, you have to be very careful about the operations on the value.  In fact, if you're not changing the value then the function by definition is "pure" since there is no change in state even if you pass the value by reference. Purity of the function doesn't talk about pointers or copy but rather the notion of state. I hope this is clear.

Anyways, I don't mind you keeping it as is as well if the other maintainers agree @raklaptudirm @siriak ??

Reply:
The go rule is "when in doubt, use a pointer". If the function is not making any changes to the input graph, a pointer is the better choice. But if it is, and we don't want it reflected in the input, we will be better off with a value.

Since this function does not modify the input graph, using a pointer is memory efficient, free from any side effects, and a good coding practice. Following the convention, I think it being a pointer is better.

Reply:
@raklaptudirm Thanks for the explanation. I've changed the function so the graph is passed as reference.

Reply:
That rule is for receivers only though, that is not generally true if we're talking about function arguments.

Reply:
@tjgurwara99  Yes, but I generally find that the rule can be used in function arguments too with good effect.